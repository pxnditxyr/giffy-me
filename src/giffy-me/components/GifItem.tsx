
interface GifItemProps {
    // id: string;
    title: string;
    url: string;
}

export const GifItem = ( { title, url } : GifItemProps ) => {

    return (
        <article
            className="flex flex-col justify-center items-center"
        >
            <img
                src={ url }
                alt={ title }
                className="w-full h-auto object-cover rounded-lg"
                />
            <p
                className="text-center text-xl font-bold text-gray-300 py-4"
            > { title } </p>
        </article>
    );
};
