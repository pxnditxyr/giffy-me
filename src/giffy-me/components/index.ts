export * from './AddNewCategory';
export * from './GifGrid';
export * from './GifItem';
