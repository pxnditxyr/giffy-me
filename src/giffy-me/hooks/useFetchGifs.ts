import { useEffect, useState } from 'react'

import { getGifsByCategory } from '../api/getGifsByCategory';
import { GifInterface } from '../interfaces'

export const useFetchGifs = ( category : string ) => {
    const [ gifs, setGifs ] = useState<Array<GifInterface>>([])
    const [ isLoading, setIsLoading ] = useState( true );

    const getGifs = async () => {
        const gifsByCategory = await getGifsByCategory( category );
        setGifs( gifsByCategory );
        setIsLoading( false );
    };

    useEffect( () => {
        getGifs();
    }, [] );

    return {
        gifs,
        isLoading
    };
};

