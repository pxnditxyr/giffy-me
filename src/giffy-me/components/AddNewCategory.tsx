import { FormEvent, useState } from 'react';
import { useForm } from '../hooks/useForm';

interface AddNewCategoryProps {
    currentCategories: Array<string>;
    onAddNewCategory: ( category : string ) => void;
}

export const AddNewCategory = ( { currentCategories, onAddNewCategory } : AddNewCategoryProps ) => {

    const [ isSame, setIsSame ] = useState<Boolean>( false );
    
    const { category, handleInputChange, handleReset } = useForm({
        category: ''
    });

    const handleSubmit = ( event : FormEvent<HTMLFormElement> ) => {
        event.preventDefault();
        
        const newCategory = category.trim();

        if ( newCategory.length < 2 ) return;

        if ( currentCategories.map( currentCategory => currentCategory.toLowerCase() ).includes( newCategory.toLowerCase() ) ) {
            setIsSame( true );
            return;
        }

        if ( isSame ) setIsSame( false );

        onAddNewCategory( newCategory );
        handleReset();
    };

    return (
        <form
            onSubmit={ handleSubmit }
            >
            <label> What are you thinking </label>
            <input
                type="text"
                name="category"
                value={ category }
                onChange={ handleInputChange }
                />
            {
                isSame 
                    && 
                ( <p> You already have this category </p> )
            }
            <button type="button"> Search </button>
        </form>
    );
};
