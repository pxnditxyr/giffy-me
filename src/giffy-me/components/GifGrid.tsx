import { GifItem } from './GifItem';

import { useFetchGifs } from '../hooks/useFetchGifs';

interface GifGridProps {
    category: string;
}

export const GifGrid = ( { category } : GifGridProps ) => {

    const { gifs, isLoading } = useFetchGifs( category );

    return (
        <div
            className="w-full flex flex-col justify-center items-center px-5"
        >
            <h3 
                className="text-center text-4xl font-bold text-amber-400 py-4 md:py-20"
            > { category } </h3>
            {
                isLoading
                    ? ( <p> Loading... </p> )
                    : (
                        <div
                            className="grid gap-4 grid-cols-1 md:grid-cols-2 lg:grid-cols-3 px-4"
                        >
                            {
                                gifs.map( ( gif ) => (
                                    <GifItem 
                                        key={ gif.id }
                                        { ...gif }
                                    />
                                ))
                            }
                        </div>
                    )
            }
        </div>
    );
};
