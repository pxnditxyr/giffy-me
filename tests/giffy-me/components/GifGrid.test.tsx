import { render, screen } from '@testing-library/react';
import { GifGrid } from '../../../src/giffy-me/components/GifGrid';

import { useFetchGifs } from '../../../src/giffy-me/hooks/useFetchGifs';

jest.mock( '../../../src/giffy-me/hooks/useFetchGifs' );

describe( 'tests in GifGrid component', () => {
    const category = 'Naruto';
    test( 'should render GifGrid component', () => {
        ( useFetchGifs as jest.Mock ).mockReturnValue({
            gifs: [],
            isLoading: true
        });
        const { container } = render( <GifGrid category={ category } /> );
        expect( container ).toMatchSnapshot();
    })

    test( 'should show loading', () => {
        ( useFetchGifs as jest.Mock ).mockReturnValue({
            gifs: [],
            isLoading: true
        });
        render( <GifGrid category={ category } /> )
        expect( screen.getByText( 'Loading...' ) );
        expect( screen.getByText( category ) );
    })

    test( 'should show items when images is charged', () => {
        const gifs = [
            {
                id: '1',
                title: 'Naruto vs Pain',
                url: 'https://thumbs.gfycat.com/ComplexSecondhandBandicoot-size_restricted.gif'
            },
            {
                id: '2',
                title: 'Maito Gai vs Madara',
                url: 'https://c.tenor.com/rWciirvM7sgAAAAd/madara-uchiha-might-guy.gif'
            }
        ];
        ( useFetchGifs as jest.Mock ).mockReturnValue({
            gifs,
            isLoading: false
        });
        render( <GifGrid category={ category } /> );
        expect( screen.getAllByRole( 'img' ).length ).toBe( 2 );
    })
})
