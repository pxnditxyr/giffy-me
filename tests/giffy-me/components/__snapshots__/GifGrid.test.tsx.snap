// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tests in GifGrid component should render GifGrid component 1`] = `
<div>
  <div
    class="w-full flex flex-col justify-center items-center px-5"
  >
    <h3
      class="text-center text-4xl font-bold text-amber-400 py-4 md:py-20"
    >
       
      Naruto
       
    </h3>
    <p>
       Loading... 
    </p>
  </div>
</div>
`;
